docs/apis/core/customfields/index.md

```diff
@@ -0,0 +1,168 @@
+
+New plugin type "customfield" was also added as part of the Custom fields API. Additional types of custom fields can be installed into /customfield/field/ .
+
+## How to use custom fields
+
+Component/plugin that uses custom fields must define a **handler class** for each area and a **configuration page**. Handler class must be called `<PLUGINNAME>/customfield/<AREA>_handler` and be placed in autoloaded location  `<PLUGINDIR>/classes/customfield/<AREA>_handler.php`. This class must extend **\core_customfield\handler** . Configuration page may be located anywhere. For course custom fields configuration the admin settings page is used [/course/customfield.php](https://github.com/moodle/moodle/blob/master/course/customfield.php). If the area uses itemid this page should take itemid as a parameter.
+
+Handler has protected constructor, to get a handler call create() method. Some areas may choose to return a singleton here:
+
+```php
+$handler = HANDLERCLASS::create($itemid);
+```
+
+Configuration page contents will be:
+
+```php
+$output = $PAGE->get_renderer('core_customfield');
+$outputpage = new \core_customfield\output\management($handler);
+echo $output->render($outputpage);
+```
+
+Handler must implement all abstract methods (calculate configuration or instance context, check permissions to configure, view or edit) and also may choose to overwrite:
+
+```php
+handler::uses_categories()
+handler::generate_category_name()
+handler::config_form_definition() // For example, the course_handler adds "locked" and "visibility" settings that control who can edit or view the particular field.
+handler::setup_edit_page() // Sets page context/url/breadcrumb for the customfield/edit.php page, in some cases it must be overridden.
+```
+
+### Add custom fields to the instance edit form
+
+Custom fields are added to the **instances**. For example, course custom fields are added to the courses, so courseid is the instanceid. In the example of [mod_surveybuilder](https://github.com/marinaglancy/moodle-mod_surveybuilder) we use $USER->id as the instanceid (which means that in this example one user can fill the survey in one module only once). In each case of using custom fields there should be a clear concept of an **instance**. Instanceid is required to save the data but it may be empty when we render the instance edit form (for example, the course is not yet created).
```

Comment:
```suggestion
Custom fields are added to the **instances**. For example, course custom fields are added to the courses, so `courseid` is the `instanceid`. In the example of [mod_surveybuilder](https://github.com/marinaglancy/moodle-mod_surveybuilder) we use `$USER->id` as the `instanceid` (which means that in this example one user can fill the survey in one module only once). In each case of using custom fields there should be a clear concept of an **instance**. `Instanceid` is required to save the data but it may be empty when we render the instance edit form (for example, the course is not yet created).
```